libcxx/test/std/utilities/optional/optional.iterator/end.pass.cpp

```diff
@@ -0,0 +1,73 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// REQUIRES: std-at-least-c++26
+
+// <optional>
+
+// constexpr iterator optional::end() noexcept;
+// constexpr const_iterator optional::end() noexcept;
+
+#include <cassert>
+#include <iterator>
+#include <ranges>
+#include <optional>
+
+template <typename T>
+constexpr bool test() {
+  std::optional<T> unengaged{std::nullopt};
+  constexpr std::optional<T> unengaged2{std::nullopt};
+
+  { // end() is marked noexcept
+    assert(noexcept(unengaged.end()));
+    assert(noexcept(unengaged2.end()));
+  }
+
+  { // end() == begin() and end() == end() if the optional is unengaged
+    auto it  = unengaged.end();
+    auto it2 = unengaged2.end();
+
+    assert(it == unengaged.begin());
+    assert(unengaged.begin() == it);
+    assert(it == unengaged.end());
+
+    assert(it2 == unengaged2.begin());
+    assert(unengaged2.begin() == it2);
+    assert(it2 == unengaged2.end());
+  }
+
+  std::optional<T> engaged{T{}};
+  constexpr std::optional<T> engaged2{T{}};
+
+  { // end() != begin() if the optional is engaged
+    auto it  = engaged.end();
+    auto it2 = engaged2.end();
+
+    assert(it != engaged.begin());
+    assert(engaged.begin() != it);
+
+    assert(it2 != engaged2.begin());
+    assert(engaged2.begin() != it2);
+  }
+
+  return true;
+}
+
+constexpr bool tests() {
+  assert(test<int>());
+  assert(test<char>());
+  assert(test<const int>());
+  assert(test<const char>());
```

Comment:
I think I like how you organized the tests now. FYI only - I am not asking for any changes: there is an alternative to iterate over multiple types a helper `types::for_each(types::integer_types(), IntegerTypesTest{});` an example usage is here: https://github.com/llvm/llvm-project/blob/9ae7490c6251ec60ab978bb818e89ca586e0c9c9/libcxx/test/std/ranges/range.factories/range.iota.view/indices.pass.cpp